File: fullSystem.cpp
#include <iostream>
#include <fstream>
#include <wiringPi.h>

// LCD - TESTING
#include "newLcdCode/CLP_lcd.h"
#include <lcd.h>
#include <vector>

#include <string>
#include <sstream>

// i2C
#include "pi2c.h";
#include <cassert>;
#include <unistd.h>;
#include <iostream>;
#include "stdlib.h";

using namespace std;

/* Files needed for compilation:
 - fullSystem.cpp
 - lcd.cpp
 - plotGraph.cpp
 - kiss_fft.h
 -  
 - "option -lwiringPi"
 - "option -lwiringPiDev"

 Compile with: g++ fullSystem.cpp pi2c.cpp -lwiringPi -lwiringPiDev -o fullSystem
*/


////////// Globals ////////////
int recordTimeSec = 1;

const int startButtonPin = 27;
const int upButtonPin = 28;
const int downButtonPin = 29;

const int teensyAddress = 7;
const int sentBytes = 128;

int lcd;



// function declarations
void waitForStart();
void updateWaitingScreen();
void setupPins();
string int2str(int num);



int main() {
	/////////////////////////////////// ON BOOT ///////////////////////////////////////
	// setup GPIO inputs and LCD
	wiringPiSetup();
	setupPins();
    lcd = setupLCD();

	// Display title text
	lcdClear(lcd);
	lcdPrint(lcd,"","Clear Lung Project","","");
	delay(2000);

    // Open output files for each mic
	ofstream mic1;
	mic1.open("mic1.txt");
	ofstream mic2;
	mic2.open("mic2.txt");

	// Display option to select recording time and wait for start button press 
	waitForStart();



	/////////////////////////////////// ON START ///////////////////////////////////////
	cout << "Recording now" << endl;
	lcdClear(lcd);
	lcdPrint(lcd,"Recording now","","","");

    // setup teensy communication
	Pi2c teensy1(teensyAddress);
	
    // send teensy the recording time (teensy starts recording)
    int recordTimeMS = recordTimeSec * 1000;
//    teensy1.i2cWrite(int2str(recordTimeMS).cstr(), (int2str(recordTimeMS).length()+1)); 

    // wait slightly longer than recording time
	delay(1000*recordTimeSec + 500);

	/*
	char recieve[chunkSize]; //temporary location to hold recieved bytes
    // Loop for reading multiple chunks of data
    for (int j=0; j<4; j++) {
        // Read a chunk from i2c
        cout << "Reading data" << std::endl;
        teensy1.i2cRead(recieve, chunkSize);
        // Print data to file
        cout << "Printing to file" << endl;
        for (int i = 0; i < chunkSize; i += 2) { 
            unsigned char lower = recieve[i];
            unsigned char upper = recieve[i+1];
            unsigned short value = (upper << 8) | lower;
            os << value << " " << endl;
        }
        // delay to let teensy read data
        usleep(600);
    }
    os.close();
	*/

/*
	// calculate total number of bytes for the recording time
	int numMics = 2;
	int totalBytes = 44100 * recordTimeSec * 2 * numMics;

	// request the total num bytes from teensy in increments of (128)
	// separate data into the files, write to files
	int mic1Counter = 0;
	int mic2Counter = 0;
	for (int j=0; j< totalBytes/chunkSize; j++) { // may need to adjust the upper loop limit
		receiveData(mic1, teensy1, chunkSize, &mic1Counter);
        usleep(600); // delay to let teensy read data
        receiveData(mic2, teensy1, chunkSize, &mic2Counter);
        usleep(600); // delay to let teensy read data
	}
	mic1.close();
	mic2.close();



	/////////////////////////////////// ON FINISH ///////////////////////////////////////
	lcdClear(lcd);
	lcdPrint(lcd,"Finished Recording","","","");

	/*
	// show graph of 1 sec of recording
	makeGraph("mic1.txt");

	// show graph of fft
	runFFT(nearestPow2(valueCounter), "mic2.txt", "fft.txt");
	makeGraph("fft.txt");
	*/

	return 0;
}




void setupPins() {
	pinMode(startButtonPin, INPUT);
	pinMode(upButtonPin, INPUT);
	pinMode(downButtonPin, INPUT);
	pullUpDnControl(startButtonPin, PUD_DOWN);
	pullUpDnControl(upButtonPin, PUD_DOWN);
	pullUpDnControl(downButtonPin, PUD_DOWN);
}


void waitForStart() {
	updateWaitingScreen();
	// while the start button hasn't been pressed
	while(!digitalRead(startButtonPin)) {
		if (digitalRead(upButtonPin)) {
			// increase the record time
			recordTimeSec++;
			cout << recordTimeSec << endl;
			updateWaitingScreen();
			delay(200);
		} else if (digitalRead(downButtonPin)) {
			// descrease record time
			if (recordTimeSec > 1) {
				recordTimeSec--;
			}
			cout << recordTimeSec << endl;
			updateWaitingScreen();
			delay(200);
		}
		
	}
}

void updateWaitingScreen() {
	string lengthString = int2str(recordTimeSec) + " sec";
	lcdClear(lcd);
	lcdPrint(lcd,"Record for", lengthString.c_str(), "Press start", "");
}

string int2str(int num) {
	stringstream strs;
	strs << num;
	return strs.str();
}

/*
// function reads in data from a teensy and writes to a file
void receiveData(ifstream &outfile, Pi2c &teensy, int bufSize, int *valueCounter) {
	char recieve[bufSize]; 
	teensy.i2cRead(recieve, bufSize);
	// convert from bytes to integer, write to file
	for (int i=0; i < bufSize; i+=2) {
		unsigned char lower = recieve[i];
		unsigned char upper = recieve[i+1];
		unsigned short value = (upper << 8) | lower;
		if (value > 10000){
			// subtract 2^16
			value -= 65536; 
		}
		outfile << *valueCounter << " " value << endl;
		(*valueCounter)++;
	}
}

*/

File: fullSystem.py

from subprocess import call
import RPi.GPIO as GPIO
import time
import serial



# Variables
recordTimeSec = 1
N = 512

startButtonPin = 16
upButtonPin = 20
downButtonPin = 21
teensyControlPin = 6

#/////////////////////////////////// FUNCTIONS ///////////////////////////////////////

# function to read state of button
def readButton(pin):
	return GPIO.input(pin)

# function to display text on lcd
def lcdprint(l1, l2, l3, l4):
	call(["./newLcdCode/cmdlineLCD",l1,l2,l3,l4])


# function to delay "t" milliseconds
def delay(t):
	time.sleep(t/1000)



# function updates lcd/record time before start
def waitForStart():
	updateWaitingScreen();
	# while the start button hasn't been pressed
	while not readButton(startButtonPin):
		global recordTimeSec
		if readButton(upButtonPin):
			# increase the record time
			recordTimeSec += 1
			print(recordTimeSec)
			updateWaitingScreen()
			delay(200);
		elif readButton(downButtonPin):
			# decrease record time
			if recordTimeSec > 1:
				recordTimeSec -= 1
			print(recordTimeSec)
			updateWaitingScreen()
			delay(200)

def updateWaitingScreen():
	lcdprint("Record for", str(recordTimeSec), "Press start", " ")



# function converts the byte data input to list of numbers
def formatDataIn(data):
	outdata = 0
	for i in range(0,length(data),2):
		lower = data[i]
		upper = data[i+1]
		value = (upper << 8) | lower
		if value > 10000:
			# subtract 2^15
			value -= 32768 #65536 
		outdata.append(value)


def makeGraph(fname):
	cmd1 = "gnuplot -e 'set terminal png; plot "
	cmd2 = " with lines' > \"graph.png\""
	# make the graph using gnuplot
	call(cmd1 + fname + cmd2)
	# open the image (install with: sudo apt-get -y install fbi)
	call("fbi -a graph.png")


def changeTeensyRecordStatus(isRecording):
	if (isRecording):
		GPIO.output(teensyControlPin, GPIO.HIGH)
	else:
		GPIO.output(teensyControlPin, GPIO.LOW)




#/////////////////////////////////// ON BOOT ///////////////////////////////////////
# setup pins
# GPIO.cleanup()
GPIO.setwarnings(False)
GPIO.setmode(GPIO.BCM)
GPIO.setup(startButtonPin, GPIO.IN, pull_up_down=GPIO.PUD_DOWN)
GPIO.setup(upButtonPin, GPIO.IN, pull_up_down=GPIO.PUD_DOWN)
GPIO.setup(downButtonPin, GPIO.IN, pull_up_down=GPIO.PUD_DOWN)
GPIO.setup(teensyControlPin, GPIO.OUT)
changeTeensyRecordStatus(False)

# show title text
lcdprint(" ","Clear Lung Project"," "," ")
delay(1000);

# open output files for each mic
mic1 = open("mic1.txt","w") 
mic2 = open("mic2.txt","w") 

# # setup the serial connection
ser = serial.Serial('/dev/ttyS0', 1000000)
ser.flushInput()
print("serial connected")

# # display option to select recording time and wait for start button press 
waitForStart()



#/////////////////////////////////// ON START ///////////////////////////////////////
print("Recording now")
lcdprint("Recording now"," "," "," ")
changeTeensyRecordStatus(True)

# wait slightly longer than recording time
delay(1000*recordTimeSec + 1500);
changeTeensyRecordStatus(False)
print("Stopped Recording")






# calculate total number of bytes for the recording time
numMics = 4
totalBytes = 44100 * recordTimeSec * 2 * numMics

# request the total num bytes from teensy in increments of (128)
# separate data into the files, write to files
mic1Counter = 0
mic2Counter = 0
bytesRead = 0


#//////////////////////////////////////////////////////////////////////////
#//////////////////////////////////////////////////////////////////////////

start = time.time()
lastReceive = start;
count = 0
while ((lastReceive - start) < 0.5):
    if ser.inWaiting >= N: #(ser.in_waiting)
        response = ser.read(N) #N)
        count = count + 1
        lastReceive = time.time()
        print(count)
        print(time.time()-start)
        ser.write('receive'.encode('utf-8'))
        mic1.write(response)
#//////////////////////////////////////////////////////////////////////////
#//////////////////////////////////////////////////////////////////////////


# # close files
mic1.close()
mic2.close()




#/////////////////////////////////// ON FINISH ///////////////////////////////////////
lcdprint("Finished Recording"," "," "," ")

# # // show graph of 1 sec of recording
# makeGraph("mic1.txt")

# # // show graph of fft
# # runFFT(nearestPow2(valueCounter), "mic2.txt", "fft.txt");
# # makeGraph("fft.txt");

# print("all done")















